### PR TITLE
Add shift (wip)

### PR DESCRIPTION
**Pull request by tom-knoetze**

### CODE REVIEW

File: player/events/events_test.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package events
+
+import (
+	"testing"
+
+	"github.com/luno/reflex/rsql"
+
+	"bitx/console/ticketing/internal/db"
```
Comment (adamhicks) [inline player/events/events_test.go]: wrong import, we will have a db/connect.go from sergio soon for db.ConnectFor...()

File: player/reflex.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+type EventType int
+
+// ReflexType returns the int representation of an EventType.
+func (t EventType) ReflexType() int {
+	return int(t)
+}
+
+// Player round event types are listed here.
+const (
+	EventTypePlayerRoundJoined     EventType = 1
```
Comment (adamhicks) [inline player/reflex.go]: default and sentinels please

File: player/reflex.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	EventTypePlayerRoundExcluded   EventType = 2
+	EventTypePlayerRoundJoining    EventType = 3
+	EventTypePlayerRoundCollecting EventType = 4
+	EventTypePlayerRoundCollected  EventType = 5
+	EventTypePlayerRoundSubmitting EventType = 6
```
Comment (adamhicks) [inline player/reflex.go]: missing Submitted?

File: player/types.go [hunk 4]
```diff
@@ -9,3 +11,46 @@ type PartInfo struct {
 	Player string
 	Part   int64
 }
+
+type PlayerRound struct {
+	ID        int64
+	Status    int64
```
Comment (adamhicks) [inline player/types.go:18]: We need the RoundId (1-10) here somewhere

File: player/types.go [hunk 5]
```diff
@@ -9,3 +11,46 @@ type PartInfo struct {
...
+	PlayerRoundStatusRoundExcluded   PlayerRoundStatus = 3
+	PlayerRoundStatusRoundJoining    PlayerRoundStatus = 1
+	PlayerRoundStatusRoundCollecting PlayerRoundStatus = 4
+	PlayerRoundStatusRoundCollected  PlayerRoundStatus = 5
+	PlayerRoundStatusRoundSubmitting PlayerRoundStatus = 6
+	PlayerRoundStatusRoundSubmitted  PlayerRoundStatus = 7
+	PlayerRoundStatusRoundEnded      PlayerRoundStatus = 8
+)
+
+// attachmentUploadReflex maps AttachmentUploadStatus types to reflex EventTypes.
```
Comment (adamhicks) [inline player/types.go]: strange naming.... :)

File: player/types.go [hunk 6]
```diff
@@ -9,3 +11,46 @@ type PartInfo struct {
...
+
+func (prs PlayerRoundStatus) Enum() int {
+	return int(prs)
+}
+
+func (aus PlayerRoundStatus) ReflexType() int {
+	return attachmentUploadReflex[aus].ReflexType()
+}
+
+func (PlayerRoundStatus) ShiftStatus() {
```
Comment (adamhicks) [inline player/types.go:33]: You can combine the statuses so we only have one enum

File: player/types.go [hunk 7]
```diff
@@ -9,3 +11,46 @@ type PartInfo struct {
...
+	PlayerRoundStatusRoundJoining    PlayerRoundStatus = 1
+	PlayerRoundStatusRoundCollecting PlayerRoundStatus = 4
+	PlayerRoundStatusRoundCollected  PlayerRoundStatus = 5
+	PlayerRoundStatusRoundSubmitting PlayerRoundStatus = 6
+	PlayerRoundStatusRoundSubmitted  PlayerRoundStatus = 7
+	PlayerRoundStatusRoundEnded      PlayerRoundStatus = 8
+)
+
+// attachmentUploadReflex maps AttachmentUploadStatus types to reflex EventTypes.
+var attachmentUploadReflex = map[PlayerRoundStatus]EventType{
```
Comment (adamhicks) [inline player/types.go]: shouldn't need this map

File: player/reflex.go (hunk 3 above)
Comment (adamhicks) [reply player/reflex.go]: actually we don't need this enum you should be able to use the other one

File: player/events/events_test.go (hunk 1 above)
Comment (tom-knoetze) [reply player/events/events_test.go]: oops

File: player/types.go (hunk 5 above)
Comment (tom-knoetze) [reply player/types.go]: lol

File: player/types.go (hunk 7 above)
Comment (tom-knoetze) [reply player/types.go]: kk

File: player/reflex.go (hunk 3 above)
Comment (tom-knoetze) [reply player/reflex.go]: ok

File: player/types.go (hunk 4 above)
Comment (tom-knoetze) [reply player/types.go:18]: Added

File: player/types.go (hunk 6 above)
Comment (tom-knoetze) [reply player/types.go:33]: done

File: player/db/rounds/rounds.go [hunk 8]
```diff
@@ -1 +1,54 @@
...
+)
+
+// Call FSM functions here.
+
+func Joining(ctx context.Context, dbc *sql.DB, roundID int64) (int64, error) {
+	return fsm.Insert(ctx, dbc, joining{RoundID: roundID})
+}
+
+func Excluded(ctx context.Context, dbc *sql.DB, id int64) error {
+	return fsm.Update(ctx, dbc, player.PlayerRoundStatusRoundJoining,
```
Comment (adamhicks) [inline player/db/rounds/rounds.go:19]: I think we need to take Tx objects here so that we can use a transaction.

**Review by adamhicks — APPROVED** [review]

**Pull request merged**